sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
     std::unique_ptr<sstable_set_impl> _impl;
-    schema_ptr _schema;
+    std::unordered_map<utils::UUID, std::unordered_set<shared_sstable>> _all_runs;
+    std::unordered_map<shared_sstable, unsigned> _times_added;
+    std::set<shared_sstable> _all;
+public:
+    versioned_sstable_set_data(std::unique_ptr<sstable_set_impl> impl);
+    void insert(shared_sstable sst);
+    void erase(shared_sstable sst);
+    std::vector<shared_sstable> select(const dht::partition_range& range) const;
```

Comment:
So, the way this works is that it gives a superset of the result (including sstables from versions other than the one we're interested in), and the caller then filters out sstables that don't belong in the version of interest?

Reply:
Correct. 
It's similar to the approach from https://github.com/scylladb/scylla/pull/7272#discussion_r494367643 - when checking if an sstable is in a certain version, we're looking for an ancestor version that added or erased that sstable, but instead of checking if the ancestor version in in the vector of versions that add or erase this sstable (`vector<version_status>` from the referenced comment), we're checking if the sstable is in the set of changed sstables in the ancestor version.
This allows us to have just shared_sstables in the base version, and this makes it easier (and more efficient) to merge versions.
On another note, we could filter the sstables before creating the entire superset by passing the version as an argument. It would take the same time, but use less memory, but I'm not sure if that's needed